File: study/开发/StockSortFilterProxyModel.py
#!/usr/bin/env python3

from PyQt5.QtCore import QSortFilterProxyModel, QRegExp


class StockSortFilterProxyModel(QSortFilterProxyModel):
    sortColumn = 0

    def __init__(self, parent=None):
        super(StockSortFilterProxyModel, self).__init__(parent)

    def SetFilterContext(self, pattern):
        self.sortColumn = self.columnCount()
        regExp = QRegExp("*%s*" % (pattern))
        regExp.setPatternSyntax(QRegExp.Wildcard)
        self.setFilterRegExp(regExp)
        print("func:SetFilterContext:heeeeeeeeeeeeeeee")

    def lessThan(self, left, right):
        return QSortFilterProxyModel.lessThan(self, left, right);

    def filterAcceptsRow(self, source_row, source_parent):
        regExp = self.filterRegExp()
        result = False
        for i in range(self.sortColumn):
            print(i)
            index = self.sourceModel().index(source_row, i, source_parent)
            print(self.sourceModel().data(index))
            context = str(self.sourceModel().data(index))
            print(context)
            if regExp.isEmpty() == False:
                regExpStr = regExp.pattern()
                result = regExp.exactMatch(context);
            if result:
                break

        return result


regExp = QRegExp("*%s*" % ("1"))
print(regExp)

File: https__downloadcsdnnet_download_qq_30392343_9673716/stocklist/sdasdasd/StockSortFilterProxyModel.cpp
#include "stdafx.h"

#include "StockSortFilterProxyModel.h"

StockSortFilterProxyModel::StockSortFilterProxyModel(QObject *parent)
	: QSortFilterProxyModel(parent)
{
	
}

StockSortFilterProxyModel::~StockSortFilterProxyModel()
{
	
}

void StockSortFilterProxyModel::SetFilterContext(const QString & pattern)
{
	sortColumn = columnCount();

    QRegExp regExp(QString("*%1*").arg(pattern));
    regExp.setPatternSyntax(QRegExp::Wildcard);
	setFilterRegExp(regExp);
}

bool StockSortFilterProxyModel::lessThan(const QModelIndex &left
									  , const QModelIndex &right) const
{
	return QSortFilterProxyModel::lessThan(left, right);
}

bool StockSortFilterProxyModel::filterAcceptsRow(int source_row
											  , const QModelIndex & source_parent) const
{
	QRegExp regExp = filterRegExp();
	bool result = false;
	for (int i = 0; i < sortColumn; ++i)
	{
		QModelIndex index = sourceModel()->index(source_row, i, source_parent);
		QString context = sourceModel()->data(index).toString();

		if (regExp.isEmpty() == false)
		{
			QString regExpStr = regExp.pattern();
			result = regExp.exactMatch(context);
		}

		if (result)
		{
			break;
		}
	}

	return result;
}
